### PR TITLE
feat: Limit of members that can be added using the autojoin feature in a team's channel to the value of the API_Users_Limit setting

### PR DESCRIPTION
**Pull request by Gustrb**
## Proposed changes (including videos or screenshots)
There existed an implicit limit of 50 members that would be included whenever a a channel with auto join was created inside a team, this task ties the limit to the API_Users_Limit setting, so it is configurable.

## Steps to test or reproduce

This task requires a team with over 1000 members, so all of these members should be created and associated with the team
1. Create a channel from the team
2. Open the kebab-menu and mark the auto join box
All the members should've been added correctly
3. Add one more user to the team
4. Create a new channel in the team
5. Mark the auto join box
Now the new user should not have been added since there is a hard limit of 1000 members

## Further comments

[SUP-34]

This PR is a copy of https://github.com/RocketChat/Rocket.Chat/pull/31423
I created this one, because the other one was created from a fork which was causing a few problems such as weird conflicts

[SUP-34]: https://rocketchat.atlassian.net/browse/SUP-34?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Gustrb** [comment]
This PR is a copy of https://github.com/RocketChat/Rocket.Chat/pull/31423
I did this because the other PR was made from a fork and was causing a few issues since the develop branch was messed up

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]
Is it possible to add tests to cover this feature?

We can use a small number for the setting to avoid creating too many users.

The ideal would be to test every case, with more and less members than the accepted, etc.

We can also add some UI tests to make sure the error is being displayed if the limit was exceeded.

File: apps/meteor/server/services/team/service.ts [hunk 1]
```diff
@@ -472,11 +473,22 @@ export class TeamService extends ServiceClassInternal implements ITeamService {
...
-			const teamMembers = await this.members(uid, room.teamId, true, undefined, undefined);
+		if (isDefault) {
+			const maxNumberOfAutoJoinMembers = settings.get<number>('API_User_Limit');
+			const teamMembers = await this.members(
+				uid,
+				room.teamId,
+				true,
+				{ offset: 0, count: maxNumberOfAutoJoinMembers },
+				// We should not get the owner of the room, since he is already a member
+				{ _id: { $ne: room.u._id } },
```
[the remark this answers is not in the document]
Comment (Gustrb) [reply apps/meteor/server/services/team/service.ts:484]: I'm not sure what you meant by this. In the code Im filtering the ids that are not equal to the owner of the room, it does not really matter which user is executing the action

File: apps/meteor/server/services/team/service.ts [hunk 2]
```diff
@@ -472,11 +473,22 @@ export class TeamService extends ServiceClassInternal implements ITeamService {
...
+			);
 
 			for await (const m of teamMembers.records) {
 				await addUserToRoom(room._id, m.user, user);
+				if (await addUserToRoom(room._id, m.user, user)) {
```
[the remark this answers is not in the document]
Comment (Gustrb) [reply apps/meteor/server/services/team/service.ts:489]: oh yea, copy paste mistake from the old PR

File: apps/meteor/server/services/team/service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarcosSpessatto) [reply apps/meteor/server/services/team/service.ts:484]: That's correct, I didn't see you are providing `room.u._Id`. Makes total sense 👍

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 3]
```diff
@@ -1570,6 +1572,85 @@ describe('[Teams]', () => {
...
+						name: teamName,
+						type: 0,
+					})
+					.expect((res) => {
+						expect(res.body).to.have.property('success', true);
+						testTeam = res.body.tea;
+					});
+
+				const testRoom = `test-room-name${Date.now()}`;
+				const createRoomPromise = request.post(api('channels.create')).set(credentials).send({
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1595]: We can use the createRoom helper function here.

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 4]
```diff
@@ -1570,6 +1572,85 @@ describe('[Teams]', () => {
...
 			});
 		});
+
+		describe('team auto-join', () => {
+			let testTeam;
+			let createdRoom;
+
+			beforeEach(async () => {
+				const teamName = `test-team-name${Date.now()}`;
+				const createTeamPromise = request
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1582]: There's a createTeam helper function within the helpers functions.

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 5]
```diff
@@ -1570,6 +1572,85 @@ describe('[Teams]', () => {
...
+						rooms: [createdRoom.body.channel._id],
+						teamName: testTeam.body.team.name,
+					});
+			});
+
+			afterEach(() =>
+				Promise.all([deleteTeam(credentials, testTeam.body.team.name), deleteRoom({ roomId: createdRoom.body.channel._id, type: 'c' })]),
+			);
+
+			it('should add API_User_Limit members when we update a team channel to be auto-join', async () => {
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1618]:
```suggestion
			it('should add members when the members count is less than or equal to the API_User_Limit setting and the team channel is auto-join', async () => {
```

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 6]
```diff
@@ -1570,6 +1572,85 @@ describe('[Teams]', () => {
...
+						isDefault: true,
+					})
+					.expect(200)
+					.expect((res) => {
+						expect(res.body).to.have.property('success', true);
+						expect(res.body).to.have.nested.property('room.usersCount').and.to.be.equal(3);
+					});
+			});
+
+			it('should not add all members when we update a team channel to be auto-join and the API_User_Limit is too small', async () => {
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1636]:
```suggestion
			it('should not add all members when we update a team channel to be auto-join and the members count is greater than the API_User_Limit setting', async () => {
```

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 7]
```diff
@@ -1765,11 +1844,18 @@ describe('[Teams]', () => {
...
+					.set(credentials)
+					.send({
+						roomId: createdRoom.body.channel._id,
+						isDefault: true,
+					})
+					.expect(200)
+					.expect((res) => {
+						expect(res.body).to.have.property('success', true);
+						expect(res.body).to.have.nested.property('room.usersCount').and.to.be.equal(2);
+					});
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1858]: We don't need to change this assertion/test case, we are already testing this behavior. Let's revert  this

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 8]
```diff
@@ -1712,17 +1793,15 @@ describe('[Teams]', () => {
...
 
-				createdRoom = await request
-					.post(api('channels.create'))
-					.set(credentials)
-					.send({
-						name: `${Date.now()}-testTeam3`,
-					});
+				createdRoom = await request.post(api('channels.create')).set(credentials).send({
+					name: testTeam3Name,
+				});
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1804]: Let's keep these unrelated changes to other specific PR.

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 9]
```diff
@@ -1570,6 +1572,85 @@ describe('[Teams]', () => {
...
+					})
+					.expect(200)
+					.expect((res) => {
+						expect(res.body).to.have.property('success', true);
+						expect(res.body).to.have.nested.property('room.usersCount').and.to.be.equal(3);
+					});
+			});
+
+			it('should not add all members when we update a team channel to be auto-join and the API_User_Limit is too small', async () => {
+				await updateSetting('API_User_Limit', 1);
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1637]: Please set the value back to defaults when we finish the tests (`after`).

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 10]
```diff
@@ -1570,6 +1572,71 @@ describe('[Teams]', () => {
...
+				testTeam = testTeamCreationResult;
+				createdRoom = testRoomCreationResult;
+
+				await request
+					.post(api('teams.addRooms'))
+					.set(credentials)
+					.expect(200)
+					.send({
+						rooms: [createdRoom.body.channel._id],
+						teamName: testTeam.body.team.name,
```
Comment (matheusbsilva137) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1594]:
```suggestion
						teamName: testTeam.name,
```
the `createTeam` helper already returns `body.team`, so this is not needed.

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 11]
```diff
@@ -1570,6 +1572,71 @@ describe('[Teams]', () => {
...
+					});
+			});
+
+			afterEach(() =>
+				Promise.all([deleteTeam(credentials, testTeam.body.team.name), deleteRoom({ roomId: createdRoom.body.channel._id, type: 'c' })]),
```
Comment (matheusbsilva137) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1599]:
```suggestion
				Promise.all([deleteTeam(credentials, testTeam.name), deleteRoom({ roomId: createdRoom.body.channel._id, type: 'c' })]),
```
Same here. This should be updated since you're using the helper

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 12]
```diff
@@ -1570,6 +1572,71 @@ describe('[Teams]', () => {
...
+					});
+			});
+
+			afterEach(() =>
+				Promise.all([deleteTeam(credentials, testTeam.name), deleteRoom({ roomId: createdRoom.body.channel._id, type: 'c' })]),
+			);
+
+			after(() => updateSetting('API_User_Limit', 250));
+
+			it('should add members when the members count is less than or equal to the API_User_Limit setting and the team channel is', async () => {
```
Comment (matheusbsilva137) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1604]:
```suggestion
			it('should add members when the members count is less than or equal to the API_User_Limit setting', async () => {
```

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]

**Review by matheusbsilva137 — APPROVED** [review]

File: .changeset/clean-moose-cover.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+---
+"@rocket.chat/meteor": minor
+"@rocket.chat/i18n": minor
+---
+
+This feature introduces the use of the API_User_Limit whenever we set a channel as to be autojoined in a team, so it follows a config value insted of a hard-coded value of 50 (which could look like a bug)
```
Comment (matheusbsilva137) [inline .changeset/clean-moose-cover.md:6]:
```suggestion
Introduced the use of the `API_User_Limit`setting to limit amount of members to simultaneoulsy auto-join a room in a team
```
Sorry for nitpicking, but we usually write changesets in the past tense and keep them small (just so that we have a pattern for the release notes). We have a guide [here](https://developer.rocket.chat/contribute-to-rocket.chat/modes-of-contribution/participate-in-rocket.chat-development/development-workflow#4.-adding-changeset-to-your-pull-request)

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]

**Review by matheusbsilva137 — APPROVED** [review]

File: apps/meteor/tests/end-to-end/api/25-teams.js [hunk 14]
```diff
@@ -1570,6 +1572,71 @@ describe('[Teams]', () => {
...
+						rooms: [createdRoom.body.channel._id],
+						teamName: testTeam.name,
+					});
+			});
+
+			afterEach(() =>
+				Promise.all([deleteTeam(credentials, testTeam.name), deleteRoom({ roomId: createdRoom.body.channel._id, type: 'c' })]),
+			);
+
+			after(() => updateSetting('API_User_Limit', 250));
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/end-to-end/api/25-teams.js:1602]:
```suggestion
			after(() => updateSetting('API_User_Limit', 500));
```
The default value is 500

File: apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx [hunk 15]
```diff
@@ -1,18 +1,41 @@
...
+			const { room: updatedRoom } = await updateRoomEndpoint({
 				roomId: room._id,
 				isDefault: !room.teamDefault,
 			});
-			dispatchToastMessage({ type: 'success', message: room.teamDefault ? 'channel set as non autojoin' : 'channel set as autojoin' });
+			if (updatedRoom.teamDefault) {
+				// If the number of members in the mainRoom (the team) is greater than the limit, show an info message
+				// informing that not all members will be auto-joined to the channel
+				const messageType = mainRoom.usersCount > maxNumberOfAutoJoinMembers ? 'info' : 'success';
+				const message = mainRoom.usersCount > maxNumberOfAutoJoinMembers ? 'Team_Auto-join_exceeded_user_limit' : 'Team_Auto-join_updated';
```
Comment (MarcosSpessatto) [inline apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx:28]: Can we add some UI tests to cover the feature, and cover this messages as well?

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx (hunk 15 above)
Comment (MarcosSpessatto) [reply apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx:28]: UI test will be added in a subsequent PR.

**Review by MarcosSpessatto — DISMISSED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

File: .changeset/clean-moose-cover.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
+---
+"@rocket.chat/meteor": minor
+"@rocket.chat/i18n": minor
+---
+
+Introduced the use of the `API_User_Limit`setting to limit amount of members to simultaneoulsy auto-join a room in a team
```
Comment (aleksandernsilva) [inline .changeset/clean-moose-cover.md:6]:
```suggestion
Introduced the use of the `API_User_Limit` setting to limit amount of members to simultaneously auto-join a room in a team
```

**Review by aleksandernsilva — CHANGES_REQUESTED** [review]

File: apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx (hunk 15 above)
Comment (aleksandernsilva) [inline apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx:28]: Passing `mainRoom` through the tree just to get `usersCount` seems a bit wasteful to me. It almost feels like that information should be returned by the endpoint performing the action (`teams.updateRoom`) 🤔

Comment (Gustrb) [reply apps/meteor/client/views/teams/contextualBar/channels/hooks/useToggleAutoJoin.tsx:28]: This value is the number of users inside the team, we can use it to tell whether we will be able to `Auto-Join` them all if the `mainRoom.usersCount` is greater than the limit.
There is no way to that in the teams.updateRoom endpoint without having to do additional queries just to get the value of the usersCount of the team.
Me and @matheusbsilva137 thought some prop drilling was worth it to not have to query the database for that data.

**Review by MarcosSpessatto — APPROVED** [review]
